Login via OpenId

BogdanScherban opened the pull request:
1) Login via OpenId - inside authorization provider;
2) Attribute dashboard in Admin-component is variable and depend on theme settings.

fzaninotto commented on src/core/App.js:
```diff
@@ -1,4 +1,5 @@
 import React, { Component } from "react";
+import { get } from "lodash";
```
Avoid bundling the entire lodash library by requiring 

```
import get from "lodash/get";
```

instead

See https://lodash.com/, "Cherry-pick methods for smaller browserify/rollup/webpack bundles."

BogdanScherban replied:
Imported "lodash/get" everywhere where we use "lodash"

fzaninotto commented on src/core/App.js:
```diff
@@ -13,10 +14,12 @@ import customRoutes from "./routes";
 const dataProvider = customDataProvider();
-
 const plugins = corePlugins.concat(nonCorePlugins);
+const Homepage = get(themeCommonElements, 'homePage', Charts);
```
again, you should do an inversion of control here

BogdanScherban replied:
@fzaninotto 
Francois, could you please explain what do you mean in "inversion of control"?

I changed "lodash" to "lodash/get" everywhere

fzaninotto replied:
instead of asking another function to return a component, consider that this App component will be rendered by other (specific) components. It's the responsibility of these components to provide the subcomponents they want to adapt the look and feel.

e.g.

```
// in packages/helm/App.js
import { App } from 'pulse-tile-core';
const HelmApp = () => <App homepage={HelmHomePage} appBar={HelmAppBar} />
export default HelmApp;

// in packages/NHS/App.js
import { App } from 'pulse-tile-core';
const NHSApp = () => <App homepage={NHSHomePage} appBar={NHSAppBar} />
export default NHSApp;
```

instead of 

```
// in packages/pulse-tile-core/App.js
const App = () => <App homepage={getHomePage()} appBar={getAppBar()} />
```

I call it an *inversion of control* because it's not up to the core component to decide which subcomponents to use anymore - this responsibility is shifted to its parent.

fzaninotto commented on src/core/dataProviders/authProvider.js:
```diff
@@ -1,32 +1,42 @@
+import jwt from "jsonwebtoken";
 import { AUTH_LOGIN, AUTH_LOGOUT, AUTH_ERROR, AUTH_CHECK, AUTH_GET_PERMISSIONS } from 'react-admin';
+
 import { token, domainName } from "../token";
```
domainName should come from "window.config"

BogdanScherban replied:
Done

fzaninotto commented on src/core/pages/InitializePage.js:
```diff
@@ -0,0 +1,273 @@
+        backgroundColor: "#fff",
+        transition: "opacity 1s ease, visibility 1s ease, background 1s ease",
+    },
+    ldsCss: {
```
strange class name

BogdanScherban replied:
Renamed

fzaninotto commented on src/core/pages/InitializePage.js:
```diff
@@ -0,0 +1,273 @@
+    },
+    slideWrapper: {
+        width: 2800,
+        WebkitAnimation: "slide 100.9s ease-out infinite",
```
And for other browsers?

BogdanScherban replied:
Added

fzaninotto commented on src/core/pages/InitializePage.js:
```diff
@@ -0,0 +1,273 @@
+    '@keyframes rings': {
+        "0%": {
+            "-webkit-transform": "rotate(0)",
+            "transform": "rotate(0)",
```
remove the quotes on the key

BogdanScherban replied:
Removed

fzaninotto commented on src/core/pages/InitializePage.js:
```diff
@@ -0,0 +1,273 @@
+                    </div>
+                </div>
+                <div className={classes.progress}>
+                    <div className={classes.progressBar} role="progressbar" style={{width: 0 + '%'}}></div>
```
Why use `style` here instead of a class?

BogdanScherban replied:
Removed

fzaninotto commented on src/core/sagas/initializeSagas.js:
```diff
@@ -0,0 +1,31 @@
+import { takeEvery, put } from 'redux-saga/effects';
+import  { get } from "lodash";
+
+import { domainName } from "../token";
```
use window.config instead

BogdanScherban replied:
I did this in previous PR.

But I did following:
1) **domainName** is imported from "../token";
2) domainName defined by condition (window && window.config....)

Can I leave this variant?
Because in this case we can:
1) don't copy _(window && window.config....)_ in each place, where we use domainName;
2) if we change name of global variable **domainName** - we will have only one place, where it be used.

fzaninotto replied:
Yes, that's all right

fzaninotto commented on src/core/sagas/initializeSagas.js:
```diff
@@ -0,0 +1,31 @@
+export default takeEvery(INITIALIZE_ACTION.REQUEST, function*(action) {
+    const url = domainName + '/api/initialise';
+    let options = {};
+    options.method = "GET";
```
not needed, GET is the default

BogdanScherban replied:
I removed this row from all requests

fzaninotto commented on src/core/sagas/initializeSagas.js:
```diff
@@ -0,0 +1,31 @@
+        options.headers = new Headers({ Accept: 'application/json' });
+    }
+    options.headers = {
+        'X-Requested-With': "XMLHttpRequest",
```
Strange. Does the API require that you pass this header?

BogdanScherban replied:
Yes, API requires it. I checked - it doesn't work without this header.

tony-shannon commented:
not sure if this PR is fit to merge @BogdanScherban

BogdanScherban commented:
@tony-shannon 
This PR can't be merged until I fix all Francois notes

BogdanScherban commented:
@fzaninotto 
Francois, I fixed all notes and updated PR. But I don't understand what do you mean here:

> again, you should do an inversion of control here

fzaninotto commented:
@tony-shannon good to merge
